### PR TITLE
[Release 3.1] | Address Shim gss api on Linux to delay loading libgssapi_krb5.so while using Net6

### PR DESCRIPTION
**Pull request by JRahnama**
Ports [dotnet/sqlclient#1411](https://github.com/dotnet/SqlClient/pull/1411) to add shim gss api on Linux following changes in [dotnet/runtime#55037](https://github.com/dotnet/runtime/pull/55037)

### Summary
When working with Kerberos authentication on Unix with recent changes in Net6 users fail to get a connection from Sql server. There is no exception thrown as well. It just crashes.

### Customer Impact
This will prevent users from upgrading to Net6 while using Kerberos authentication in Unix.

### Testing
Adding tests require some special setup on CI pipelines to install Net6 TFS and enabled Kerberos authentication.

cc: @danmoseley @saurabh500 @David-Engel

### CODE REVIEW

**Review by danmoseley — APPROVED** [review]

**Review by danmoseley — DISMISSED** [review]
I didn't notice you were putting this in 3.1. I believe this is needed in 6.0. That would be the dotnet/runtime repo, in the release/6.0 branch.

**Comment by danmoseley** [comment]
linking https://github.com/dotnet/SqlClient/issues/1390

**Comment by David-Engel** [comment]
> I didn't notice you were putting this in 3.1. I believe this is needed in 6.0. That would be the dotnet/runtime repo, in the release/6.0 branch.

@danmoseley 

I thought System.Data.SqlClient was removed from runtime after 3.1 and this was the last servicing branch that would contain it. This is where we've backported all SDS fixes in the past:
https://github.com/dotnet/corefx/pulls?q=is%3Apr+author%3Acheenamalhotra
And there are zero SDS PRs in dotnet/runtime:
https://github.com/dotnet/runtime/pulls?q=is%3Aopen+is%3Apr+label%3Aarea-System.Data.SqlClient

Service releases of SDS from this 3.1 branch rely on continued backwards compatibility in .NET 5+.

**Comment by danmoseley** [comment]
My assumption was that the fix needed to be in the libraries themselves.

It is possible I didn't think about this clearly (moving too fast). Let's discuss in https://github.com/dotnet/runtime/pull/65469

**Comment by David-Engel** [comment]
You might be right. I didn't realize the code change was not in a SqlClient-specific file. There is probably a nuance I'm missing, too.

**Comment by carlossanlop** [comment]
@David-Engel @danmoseley the branch is open for servicing changes. What should we do about this PR?

**Comment by David-Engel** [comment]
> What should we do about this PR?

System.Data.SqlClient still needs the fix. But I think the PR needs to be updated. From [jkotas' comment](https://github.com/dotnet/runtime/pull/65469#issuecomment-1043917388):

> The delta in this PR affects both the core runtime assemblies and System.Data.SqlClient. The change in core runtime assemblies is unnecessary - it comes with unnecessary risk. If you go with adding the workaround to release/3.1, it should be authored to affect System.Data.SqlClient only.

@JRahnama It sounds like the fix needs to be moved into an appropriate file that only affects System.Data.SqlClient.

**Comment by ericstj** [comment]
This PR needs changes and needs tactics approval before merging.

**Review by jkotas — APPROVED** [review]

**Comment by David-Engel** [comment]
> This PR needs changes and needs tactics approval before merging.

@ericstj @danmoseley - Changes have been made. What's involved with getting tactics approval for this? We have one PR approval from Jan. Do we need any others?

File: src/System.Data.SqlClient/src/Common/Interop/Unix/System.Net.Security.Native/Interop.NetSecurityNative.Extended.cs
Comment (ericstj) [inline src/System.Data.SqlClient/src/Common/Interop/Unix/System.Net.Security.Native/Interop.NetSecurityNative.Extended.cs:12]: Typically we don't think of the native shims as public API.  Since you're depending on this from a NuGet package it's now effectively public API that can never change.  Has that been discussed with the owners of that component? @dotnet/ncl @VSadov 

Also, I don't see this export at all in 3.1: https://github.com/dotnet/corefx/blob/release/3.1/src/Native/Unix/System.Net.Security.Native/pal_gssapi.c

Do you need to port https://github.com/dotnet/runtime/pull/55037 as well?
(edit) I see you're trying to put this behind a version check to avoid calling it in 3.1.

**Comment by ericstj** [comment]
> What's involved with getting tactics approval for this? We have one PR approval from Jan. Do we need any others?

Once it's ready for approval you need to add the `servicing-consider` label and someone needs to represent in the tactics meeting.  When you're ready, reach out to Dan or myself over email and we can share the invite.

File: src/System.Data.SqlClient/src/Common/Interop/Unix/System.Net.Security.Native/Interop.NetSecurityNative.Extended.cs
Comment (VSadov) [reply src/System.Data.SqlClient/src/Common/Interop/Unix/System.Net.Security.Native/Interop.NetSecurityNative.Extended.cs:12]: > Typically we don't think of the native shims as public API. Since you're depending on this from a NuGet package

That is correct. We treat the shims as implementation details - basically a part of the corresponding managed library that happens to be in native code. Sometimes shim functionality gets rewritten in managed code and native part is removed. Even tests avoid calling shims directly.

Is this change adding another managed client to the shim or that happened long time ago?

Comment (jkotas) [reply src/System.Data.SqlClient/src/Common/Interop/Unix/System.Net.Security.Native/Interop.NetSecurityNative.Extended.cs:12]: This is very old issue: https://github.com/dotnet/runtime/issues/24115

The networking team promised to maintain compatibility of System.Net.Security.Native shims for SqlClient as a short-term fix. That was in 2017. We are still waiting on the long-term fix.

Comment (David-Engel) [reply src/System.Data.SqlClient/src/Common/Interop/Unix/System.Net.Security.Native/Interop.NetSecurityNative.Extended.cs:12]: This is an effort to keep System.Data.SqlClient functional against .NET 6+. Fixing the internals calls by both SqlClients (dotnet/runtime#24115) is blocked by dotnet/runtime#29270, which is the public API replacement for those calls.

**Comment by carlossanlop** [comment]
@David-Engel 

- The latest changes in this PR have been signed off by @jkotas .
- Many CI legs failed. Are they related to this change?
- The PR did not have the `servicing-consider` label and still has the `* NO MERGE *` label, so it didn't go to Tactics. Please add them if it's ready to go through Tactics.

It will have to wait for the next servicing release because today is the Code Complete due date for July servicing, unless the label is added before the Tactics meeting tomorrow, and the CI question is answered.

**Comment by carlossanlop** [comment]
@David-Engel @JRahnama The branch is open once again for servicing changes. Can you please address the above items? Otherwise, if the change isn't needed anymore, please close the PR.

**Comment by David-Engel** [comment]
@carlossanlop 

> * Many CI legs failed. Are they related to this change?

I don't believe so. This change is all under SqlClient and the failures appear to be related to other areas.

> * The PR did not have the `servicing-consider` label and still has the `* NO MERGE *` label, so it didn't go to Tactics. Please add them if it's ready to go through Tactics.

Removed `* NO MERGE *` and added `servicing-consider`.

**Comment by danmoseley** [comment]
@ericstj could someone please confirm no packaging changes required here?

**Comment by danmoseley** [comment]
@bartonjs does this look good to you too?

**Comment by danmoseley** [comment]
I didn't intentionally remove @saurabh500 and now it won't let me put him back.

**Comment by danmoseley** [comment]
Note that System.Data.SqlClient built from this repo does not go out of support in Dec along with .NET 3.1 -- the sources here remain the latest bits for the package (with a high bar for servicing). Microsoft.Data.SqlClient remains recommended for new code.

@ericstj I wonder whether we should consider whether we should move this code elsewhere (perhaps runtime in the 7.0 branch) so we can fully retire corefx when we move the ASP.NET relevant bits from 2.1 also?

**Review by ericstj — CHANGES_REQUESTED** [review]
> could someone please confirm no packaging changes required here?

Packaging changes are required.  Here was a previous commit that did this: https://github.com/dotnet/corefx/commit/ad726016a345c82123141b63bb8fecfc3c887f86

**Comment by ericstj** [comment]
> @ericstj I wonder whether we should consider whether we should move this code elsewhere (perhaps runtime in the 7.0 branch) so we can fully retire corefx when we move the ASP.NET relevant bits from 2.1 also?

Some history here: we had set the precedent in 1.x that packages would be maintained in `main` so that we could keep servicing costs down.  Want an update? - just move to latest.  In 2.0 we removed many of the packages that were absorbed by the shared framework and netstandard because we knew we'd never need to ship them again - they contained no serviceable code since the implementations moved into the frameworks (.NETCore shared framework and .NETFramework).   It was later that some packages were removed that were not absorbed by frameworks - or not all frameworks.  These were removed based on the premise that they were "done" and would go out of support when their branch when out of support.  It sounds like that's not the case [here](https://github.com/dotnet/runtime/pull/2275).  What's changed since then?  Why are we changing the POR?

**Comment by jkotas** [comment]
> we move the ASP.NET relevant bits from 2.1

Where are we moving the ASP.NET relevant bits from 2.1?

I am wondering whether we need to have dotnet/runtime-graveyard repo for packages that are in a perpetual servicing-only mode and that does not really make sense to carry in dotnet/runtime.

**Comment by danmoseley** [comment]
@jkotas I was referring to discussion about moving the few bits in this repo that are supported indefinitely as part of ASP.NET 2.1 on .NET Framework into the aspnet repo soon so any servicing can happen in one place. @ericstj is the owner of this plan and can speak to it and correct me if needed.

**Review by bartonjs — APPROVED** [review]
I share in the upset that a NuGet package is depending directly on an inbox shim, but this isn't introducing the problem.

NetSecurityNative_EnsureGssInitialized looks like it is, observationally, idempotent. So, the initializer being called twice won't break things.  (It doesn't have the possibility of any fn_ptr being temporarily reassigned to nullptr that the crypto shim had at one point)

**Comment by carlossanlop** [comment]
@JRahnama @DavoudEshtehari please add the OOB package authoring changes so this can merge.

I think none of the test failures are related, but can you please also double check?

**Comment by JRahnama** [comment]
> please add the OOB package authoring changes so this can merge.

I am not sure if I know what to do to accomplish this. Can you explain a bit more please?




> I think none of the test failures are related, but can you please also double check?

The code is inside corefx and as far as I was able to see nothing was related to this change.

**Comment by carlossanlop** [comment]
@JRahnama you need to include changes that are very similar to the ones @ericstj shared above: https://github.com/dotnet/corefx/commit/ad726016a345c82123141b63bb8fecfc3c887f86

The idea is to bump the version of the package that needs to be generated for your assembly, to ensure it gets properly shipped.

**Comment by JRahnama** [comment]
@carlossanlop I added required OOB for System.Data.SqlClient.

**Comment by danmoseley** [comment]
Can someone signoff on the packaging change?

File: pkg/Microsoft.Private.PackageBaseline/packageIndex.json
Comment (carlossanlop) [inline pkg/Microsoft.Private.PackageBaseline/packageIndex.json:1555]: This wasn't changed in the old commit https://github.com/dotnet/corefx/commit/ad726016a345c82123141b63bb8fecfc3c887f86 .

According to the discussion [the old PR](https://github.com/dotnet/corefx/pull/42952/files#r455247565), the baseline wasn't expected to be bumped. @ericstj is it still the case here too? Note that the old discussion was specific to the Microsoft.Windows.Compatibility section.

File: pkg/Microsoft.Windows.Compatibility/Microsoft.Windows.Compatibility.pkgproj
Comment (carlossanlop) [inline pkg/Microsoft.Windows.Compatibility/Microsoft.Windows.Compatibility.pkgproj:33]: Ok, according to @ericstj [in the old PR](https://github.com/dotnet/corefx/pull/42952#issuecomment-658533322), it is expected to bump the version of Microsoft.Windows.Compatibility too. I assume it's also expected to do it here as well. But you're missing the changes in lines 6 and 7 of this file.

**Review by carlossanlop — COMMENTED** [review]
@JRahnama the changes are incomplete. You also need to:

- Bump the "StableVersions" number for Microsoft.Windows.Compatibility in packageIndex.json, line 430. Do _not_ bump the "BaselineVersion" under it.
- Bump the "PackageVersion" and "ServiceModelVersion" numbers in Microsoft.Windows.Compatibility.pkgproj.
- Bump "PackageVersion" and "AssemblyVersion" in src/System.Data.SqlClient/Directory.Build.props.
- Add the two project includes in src/package.builds.
  
Basically, make sure to match the same changes that were made in [the old PR](https://github.com/dotnet/corefx/pull/42952/files).

**Comment by JRahnama** [comment]
ok, thanks for pointing out the missing parts. I have updated them now.

**Comment by danmoseley** [comment]
This was probably mentioned already, but do we expect to push this out aligned with core .NET bits (presumably Nov 8th at this point) or just push out on its own? I assume the latter?

**Comment by danmoseley** [comment]
Test failures are (of course) unrelated. Some bit rot in unrelated tests, not worth fixing at this point.

**Comment by David-Engel** [comment]
> This was probably mentioned already, but do we expect to push this out aligned with core .NET bits (presumably Nov 8th at this point) or just push out on its own? I assume the latter?

We have no preference on this item. Whatever is more convenient for the servicing process.

File: pkg/Microsoft.Private.PackageBaseline/packageIndex.json
Comment (carlossanlop) [reply pkg/Microsoft.Private.PackageBaseline/packageIndex.json:1555]: @JRahnama this conversation was resolved without an answer. Did you get a chance to confirm if the change was needed or not?

Comment (carlossanlop) [reply pkg/Microsoft.Private.PackageBaseline/packageIndex.json:1555]: @ericstj @ViktorHofer if you could help confirm this, that would be extremely helpful

**Comment by carlossanlop** [comment]
@JRahnama @David-Engel there is still one more question that needs an answer. I unresolved it.

**Comment by David-Engel** [comment]
> there is still one more question that needs an answer. I unresolved it.

@carlossanlop 
We don't know the answer to the question in `pkg/Microsoft.Private.PackageBaseline/packageIndex.json`. We'll defer to @ericstj and @ViktorHofer .

File: pkg/Microsoft.Private.PackageBaseline/packageIndex.json
Comment (ericstj) [reply pkg/Microsoft.Private.PackageBaseline/packageIndex.json:1555]: BaseLineVersion is used to control the version referenced by other packages when their assemblies have a SqlClient reference (even older versions).  I believe we decided we'd update it regularly since that matches the behavior for runtime servicing.

**Review by ericstj — APPROVED** [review]

**Comment by carlossanlop** [comment]
Thanks @ericstj. Merging.

**Comment by danmoseley** [comment]
Do we need to do anything particular to ship this now, or does it "join the flow"? (I feel a moral obligation to ensure we don't trip over ourselves somehow since I am partially responsible for the delays here earlier in the year!)

**Comment by danmoseley** [comment]
@carlossanlop curious about the q above?

File: src/packages.builds
Comment (joperezr) [inline src/packages.builds:19]: These changes were made in the wrong ItemGroup, they should have instead been made in the one bellow which is the one used for servicing.

https://github.com/dotnet/corefx/blob/c6f58cb8a0946b89c0faad564bbce8a102ca7211/src/packages.builds#L27-L35

The result is that these two packages are not going to be built as part of the official build, so in order for the packages to get serviced that will need to be fixed.

**Comment by JRahnama** [comment]
@joperezr if I make a new PR right now and move those newly added projects to the below `ItemGroup` do we need to wait for the .Net Tactics to approve the changes and then move forward?

**Comment by carlossanlop** [comment]
No, it's infra only, we can take it without requesting additional Tactics approval.

**Comment by carlossanlop** [comment]
FYI 4.8.4 has been published, which includes this fix: https://www.nuget.org/packages/System.Data.SqlClient/4.8.4

**Pull request merged**
